### PR TITLE
Add SRU Test Plan document

### PR DESCRIPTION
**Pull request by mateusrodrigues**
Adding a document that describes the process of testing a .NET package being SRU'ed. This will be included as part of the SRU bug template and utilized by testers for SRU verification purposes.

This will make it easier to maintain an up-to-date test plan and link SRU bugs in Launchpad to a piece of documentation that will live longer than the bug itself.

### CODE REVIEW

File: docs/SRUTestPlan.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+With a missing NuGet package, running the project should now fail:
+```
+$ dotnet run
+```
+Expected output:
+```
+/home/ubuntu/Testing/Program.cs(6,7): error CS0246: The type or namespace name 'Google' could not be found (are you missing a using directive or an assembly reference?) [/home/ubuntu/Testing/Testing.csproj]
+
+The build failed. Fix the build errors and run again.
+```
```
Comment (dviererbe) [inline docs/SRUTestPlan.md:301]: questions: Should we get rid of these tests? I see no significant value in running these tests manually. The testsuite runs tests that cover these scenarios. If the testsuite does not cover some aspects, than we should add it to the testsuite, rather than running them manually.

note: Similar things could be said about the above section, but I think that there is some value in a human checking that all the version numbers are correct. There are a lot of edge cases that we encountered in the past.

Comment (dviererbe) [reply docs/SRUTestPlan.md:301]: todo: The test plan should contain running the testsuite. An Ubuntu autopkgtest cloud run log is sufficient.

Comment (mateusrodrigues) [reply docs/SRUTestPlan.md:301]: Well, the SRU process does describe a manual test review ([Verification](https://wiki.ubuntu.com/StableReleaseUpdates#Verification)) that's independent from autopkgtest results.

So, I think that having a manual set of steps that anyone can follow to test .NET is important, especially for SRUs that track new micro-releases. Honestly, I believe that asking verifiers to run the testsuite (effectively, the autopkgtest suite) locally doesn't add any value, since that has already run on the autopkgtest cloud.

If there's ever a .NET SRU that fixes a specific issue, such as [LP: #2040547](https://bugs.launchpad.net/ubuntu/+source/dotnet8/+bug/2040547), then we can replace this general set with one that specifically tests the bug fix being deployed.

Comment (dviererbe) [reply docs/SRUTestPlan.md:301]: Let me further expand my rationale. If you are not convinced afterwards, I will approve the PR.

Usually an Ubuntu package very rarely receive an SRU and this is reflected in the SRU process verification section. Requesting an SRU is an exceptional event in the lifecycle of an Ubuntu package and therefore requires careful consideration. A manual test plan makes sense, because humans can easily perform a wide range of coverage tests, that would take much longer to automate and the manual test plan is unlikely to be reused after an SRU.

.NET micro-releases are different from the usual SRUs, because we release .NET in predictable one month intervals. Therefore .NET packages are expected to receive multiple SRUs during it's lifetime. Combine this with the fact that humans are horrible at consistently executing simple tasks repeatedly and take much longer to do so compared to automated tests, which results in low confidence regression test results which also delay the deployment.

I would therefore argue that requiring manual tests for the .NET SRU verification is counterproductive for the goal of detecting regressions. If the automatic test suite does not give us enough confidence in the quality of the builds, than we should improve the test suite.

---

Unrelated to the above:
Maybe we could provide this as a simple-to-run community smoke test suite, for everyone who wants to contribute towards an SRU bug verification. Having people with unusual systems configurations run this tests provides additional test coverage.

Comment (mateusrodrigues) [reply docs/SRUTestPlan.md:301]: > .NET micro-releases are different from the usual SRUs, because we release .NET in predictable one month intervals. Therefore .NET packages are expected to receive multiple SRUs during it's lifetime. Combine this with the fact that humans are horrible at consistently executing simple tasks repeatedly and take much longer to do so compared to automated tests, which results in low confidence regression test results which also delay the deployment.

Yes, I agree with you. My argument is simply that this is a process requirement that has to be fulfilled regardless. The .NET packages don't yet have a documented SRU exception that allows us to simply say that our test suite is good enough so there is no need to manually verify releases -- with the exception of actual bugfix SRUs, of  course, in which case a test plan that verifies the bug fix is valuable. So, this is a great point to be made to the SRU team once we start working on getting that exception :)

> I would therefore argue that requiring manual tests for the .NET SRU verification is counterproductive for the goal of detecting regressions. If the automatic test suite does not give us enough confidence in the quality of the builds, than we should improve the test suite.

So I would say that we can drop these manual tests once we have an agreement with the SRU team that, from then on, .NET SRUs are to be handled a certain way and manual verification should not be required for micro-release SRUs given the breadth of out testsuite. Until then, this is something they'll still ask for.

Unless you believe we can actually start making that argument in SRU bugs now, in which case I'd be 100% down for dropping these tests in lieu of the testsuite.

Comment (dviererbe) [reply docs/SRUTestPlan.md:301]: > Unless you believe we can actually start making that argument in SRU bugs now

I think we can, regarding https://wiki.ubuntu.com/StableReleaseUpdates#New_upstream_microreleases

**Review by dviererbe — APPROVED** [review]
LGTM

note: We will discuss with the SRU team the extend of required manual approval. Additionally we are working on a spec to standardize the release process.
